okio/src/jvmMain/kotlin/okio/Timeout.kt

```diff
@@ -193,10 +193,19 @@ actual open class Timeout {
 
       if (waitNanos <= 0) throw InterruptedIOException("timeout")
 
+      val cancelMarkBefore = cancelMark
+
       // Attempt to wait that long. This will return early if the monitor is notified.
       val nanosRemaining = condition.awaitNanos(waitNanos)
 
-      if (nanosRemaining <= 0) throw InterruptedIOException("timeout")
+      // If there's time remaining, we probably got the call we were waiting for.
+      if (nanosRemaining > 0) return
+
+      // Return without throwing if this timeout was canceled while we were waiting. Note that this
+      // return is a 'spurious wakeup' because Condition.signal() was not called.
+      if (cancelMark !== cancelMarkBefore) return
+
+      throw InterruptedIOException("timeout")
```

Comment:
Not my favorite control flow structure, but probably is the most readable.

Reply:
Note that this throw doesn’t get caught by the directly-enclosing try/catch. InterruptedIOException and InterruptedException are tricky.

Reply:
Yeah I was going to suggest moving the throw out of the try/catch, but then I really want the returns out of the try/catch too, but then local scoping becomes hard and it's all terrible and whatever this is fine.